perma_web/perma/tasks.py

```diff
@@ -302,586 +704,331 @@ def warn_on_exception(message="Exception in block:", exception_type=Exception):
-                else:
-                    if "perma" in directive and "noarchive" in directive:
-                        darchive = True
-
-            if darchive:
-                save_fields(link, is_private=True, private_reason='policy')
-                print "x-robots-tag found, darchiving"
-
-        # get favicon urls
-        # Here we fetch everything in the page that's marked as a favicon, for archival purposes.
-        # But we only record a favicon as our favicon_capture_url if it passes a mimetype whitelist.
-        def favicon_thread():
-            favicon_urls = []
-            if have_html and browser_still_running(browser):
-                favicons = repeat_while_exception(lambda: browser.find_elements_by_css_selector('link[rel="shortcut icon"],link[rel="icon"]'),
-                                                  timeout=10)
-                for candidate_favicon in favicons:
-                    if candidate_favicon.get_attribute('href'):
-                        candidate_favicon_url = urlparse.urljoin(content_url, candidate_favicon.get_attribute('href'))
-                        favicon_urls.append(candidate_favicon_url)
-            favicon_urls.append(urlparse.urljoin(content_url, '/favicon.ico'))
-            if not favicon_urls:
-                return
+                    wait_time = time.time() - start_time
+                    if wait_time > RESOURCE_LOAD_TIMEOUT:
+                        raise HaltCaptureException
 
-            for favicon_url in favicon_urls:
-                print "Fetching favicon from %s ..." % favicon_url
-                favicon_response, e = get_url(favicon_url)
-                if e or not favicon_response or not favicon_response.ok:
-                    print "Favicon failed:", e, favicon_response
-                    continue
+                    inc_progress(wait_time/RESOURCE_LOAD_TIMEOUT, "Fetching target URL")
+                    time.sleep(1)
 
-                # apply mime type whitelist
-                mime_type = favicon_response.headers.get('content-type', '').split(';')[0]
-                if mime_type not in VALID_FAVICON_MIME_TYPES:
-                    continue
+            print "Fetching robots.txt ..."
+            run_capture_component(thread_list, robots_txt_thread, kwargs={
+                "link": link,
+                "target_url": target_url,
+                "content_url": content_url,
+                "thread_list": thread_list,
+                "proxy_address": proxy_address
+            })
+
+            inc_progress(1, "Checking x-robots-tag directives.")
+            if xrobots_blacklists_perma(robots_directives):
+                safe_save_fields(link, is_private=True, private_reason='policy')
+                print "x-robots-tag found, darchiving"
 
-                successful_favicon_urls.append((favicon_url, mime_type))
+            if have_html:
+                # DO EVERYTHING THAT'S SAFE TO DO BEFORE "ONLOAD" IS FIRED
```

Comment:
(A) Yes.... this division and accompanying comments are somewhat mendacious. Re-examining true motivations: doing meta tags here means we very reliably get NYtimes meta tags, and doing title after onload means we very reliably get accurate titles for forbes articles, rather than the title of their splash screen. But, of course, that means we are at least sometimes analyzing the meta tags on the forbes welcome page, rather than those of the Forbes article. The favicons, on the other hand, again, reanalyzing motivations.... don't matter that much, so it felt "safe" to put that thread here, to see if we could shave a few seconds off the capture process. But that's not necessarily a great decision either.

For title, description, and meta tags, the right thing to do is definitely to grab the html both immediately and after onload, and maybe move the analysis into the "finally" section, since its so zippy, so that it runs regardless of whether the browser has crashed. I'll see what that looks like.

Favicons..... I still think doing it here is fine, but I don't feel strongly about it.

B) Yes: Since we know that we've received a complete, contentful response already at this point, Selenium's implicit wait (set to a generous 2 seconds by browser.implicitly_wait(ELEMENT_DISCOVERY_TIMEOUT)) should more than take care of it. (I just double checked that it waits for execute_script and other commands, not just the explicit element finding commands.) In the case of meta tags, we also have repeat_until_truthy going, just in case.